### ISSUE TITLE
virtual destructors

### ISSUE THREAD

**Issue by tomasz-grobelny** [Kistler-Group/sdbus-cpp#112]
Is there any reason why sdbus-c++-xml2cpp generates code with non-virtual destructors?

**Comment by sangelovic** [comment]
If you look more closely, the code itself reveals why, since it applies [protected non-virtual destructor idiom](http://www.gotw.ca/publications/mill18.htm).

In short, sdbus-c++-xml2-pp-generated classes are not meant to be used as interface classes by library clients, and specifically of that, they are not meant to be deleted polymorphically. They are meant to be overridden by clients, that's all. Hence, non-virtual protected destructor idiom is applied, and the virtual methods in both proxy and adaptor generated classes are protected.

**Comment by tomasz-grobelny** [comment]
In this case would it make sense to add some pragmas to disable warnings issued by compilers? I am trying to use sdbus-cpp in anbox and came across this issue.

And another thing is that in Message.h and TypeTraits.h there is "struct UnixFd;" forward declaration while definition in Types.h is "class UnixFd". Should this be fixed?

**Comment by sangelovic** [comment]
What kind of warning are you getting there? And which compiler is it that
is giving you a warning on that virtual destructor?

If it's just a non-virtual dtor warning, then pragmas or any code changes
are not a solution at all. The code is valid, safe and idiomatic. What I
would do perhaps is stop using that compiler :) Seriously. I would expect
such warning from a dumb, sub-par linter tool (maybe it's this one that is
giving you this warning?), but surely not from a (mature) compiler.

The latter - mixing of class and struct keyword - is a typo. Thanks for
pointing to it, and please would you submit a pull request? Thaaanks.

**Comment by tomasz-grobelny** [comment]
Ah, my bad. The warning I am getting is from sdbus::AdaptorInterfaces class which is base for my class implementing dbus interface (according to tutorial). And it seems to be correctly generated by gcc 9.3.0 with -Wnon-virtual-dtor command line option.

As for typo - I will when I find some time.

**Comment by tomasz-grobelny** [comment]
Here is the excerpt from gcc output:
```
/home/tomek/anbox/external/sdbus-cpp/include/sdbus-c++/AdaptorInterfaces.h: In instantiation of 'class sdbus::AdaptorInterfaces<org::anbox::ApplicationManager_adaptor>':
/home/tomek/anbox/src/anbox/dbus/ApplicationManagerServer.h:6:48:   required from here
/home/tomek/anbox/external/sdbus-cpp/include/sdbus-c++/AdaptorInterfaces.h:91:11: warning: 'class sdbus::AdaptorInterfaces<org::anbox::ApplicationManager_adaptor>' has virtual functions and accessible non-virtual destructor [-Wnon-virtual-dtor]
   91 |     class AdaptorInterfaces
      |           ^~~~~~~~~~~~~~~~~
```

**Comment by sangelovic** [comment]
Yes, `AdaptorInterfaces` and `ProxyInterfaces` didn't have the destructor protected, so I made that explicit. I also fixed the struct/class inconsistency.

Additional note: gcc may also issue this warning on your final classes implementing `AdaptorInterfaces`/`ProxyInterfaces`. One may be tempted to just mark the destructor of these virtual. But that's not a solution -- it will only confuse readers and will have useless runtime performance cost. The solution is to mark these final class with `final` keyword, and the warning disappaers. (Because this make it impossible to have a class lower in the hierarchy where the class lower in the hierarchy could be deleted through the pointer to the upcasted pointer to this class.)
